xc/common/utils/prjxray_create_place_constraints.py

```diff
@@ -189,26 +199,39 @@ def get_tile_dict(self):
 
 
 class ClockPlacer(object):
-    def __init__(self, vpr_grid, io_locs, blif_data, roi):
+    def __init__(self, vpr_grid, io_locs, blif_data, args):
 
-        self.roi = roi
+        self.roi = args.roi
         self.cmt_to_bufg_tile = {}
         self.bufg_from_cmt = {
             'TOP': [],
             'BOT': [],
         }
-
         cmt_dict = vpr_grid.get_cmt_dict()
 
-        top_cmt_tile = next(
-            k for k, v in cmt_dict.items() if k.startswith('CLK_BUFG_TOP')
-        )
-        bot_cmt_tile = next(
-            k for k, v in cmt_dict.items() if k.startswith('CLK_BUFG_BOT')
-        )
+        try:
+            top_cmt_tile = next(
+                k for k, v in cmt_dict.items() if k.startswith('CLK_BUFG_TOP')
+            )
+            _, thresh_top_y = cmt_dict[top_cmt_tile]['canon_loc']
+        except StopIteration:
+            thresh_top_y = None
 
-        _, thresh_top_y = cmt_dict[top_cmt_tile]['canon_loc']
-        _, thresh_bot_y = cmt_dict[bot_cmt_tile]['canon_loc']
+        try:
+            bot_cmt_tile = next(
+                k for k, v in cmt_dict.items() if k.startswith('CLK_BUFG_BOT')
+            )
+            _, thresh_bot_y = cmt_dict[bot_cmt_tile]['canon_loc']
+        except StopIteration:
+            thresh_bot_y = None
+
+        if args.graph_limit is None:
+            assert thresh_top_y is not None, "CMTs in the top half of the device not found"
+            assert thresh_bot_y is not None, "CMTs in the bottom half of the device not found"
+        else:
+            assert thresh_top_y is not None or thresh_bot_y is not None, (
+                "The device grid does not contain any CMTs"
```

Comment:
A better message would be "does not contain any BUFG sites".  If now BUFG's are available, then any source to BUFG connections become effectively impossible, without something special.

Reply:
Should I also change the previous assert messages to "BUFG sites in the top/bottom half of the device not found"?

Reply:
Ya, that is more precise.